man/man1/backgrounds.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "BACKGROUNDS" "1" "December 06, 2021" "backgrounds 4.0" "User Manual"
.hy
.SH NAME
.PP
backgrounds - Manage desktop wallpapers, backgrounds, and slideshows
.SH SYNOPSIS
.PP
\f[B]backgrounds\f[R] [ \f[B]-u\f[R] ] [ \f[B]-a\f[R] ] [ \f[B]-l\f[R] ]
[ \f[B]-x\f[R] ] [ \f[B]-pP\f[R] ] [ \f[B]-S\f[R] ] [ \f[B]-n\f[R]
numpics ] [ \f[B]-s\f[R] subdir ] directory
.SH DESCRIPTION
.PP
\f[B]backgrounds\f[R] can be used to populate folders with image files
to be cycled through and used as desktop wallpaper.
When invoked as \f[I]slides\f[R] or \f[I]slideshow\f[R] it displays a
slideshow of prepared image folders.
Convenience links for \f[I]slides\f[R] and \f[I]slideshow\f[R] are
created during installation of the DoctorFreeScripts package.
.SH COMMAND LINE OPTIONS
.TP
\f[B]-u\f[R]
display usage message
.TP
\f[B]-a\f[R]
indicates add to existing background/slide pics
.TP
\f[B]-l\f[R]
lists currently installed background/slide pics
.TP
\f[B]-p\f[R]
indicates search for background pics in prepared folders
.TP
\f[B]-P\f[R]
indicates search for background pics in prepared folders and update
.TP
\f[B]-n numpics\f[R]
sets maximum number of pics to be copied (default 2048)
.TP
\f[B]-s subdir\f[R]
searches in the root directory of your images for specified folder (e.g.
on Linux systems search in \f[I]/u/pictures\f[R] for \f[I]subdir\f[R])
.TP
\f[B]-x\f[R]
indicates use XnViewMP for slideshow
.TP
\f[B]-S\f[R]
indicates run a slideshow of pics
.SH EXAMPLES
.TP
\f[B]backgrounds -l\f[R]
List any installed wallpaper images in
\f[B]/usr/local/share/backgrounds/\f[R] and indicate the location of
prepared background image folders
.TP
\f[B]slideshow\f[R]
Display a slideshow of images from the background images folder
.TP
\f[B]backgrounds -p Owls\f[R]
Searches the prepared image folders for a folder named \f[I]Owls\f[R]
and copies any images in that folder into the default backgrounds folder
for use as desktop wallpapers and slideshows.
Any previous images in the default backgrounds folder are removed and
replaced with the images from the \f[I]Owls\f[R] folder.
.TP
\f[B]backgrounds -a -p Fractals\f[R]
Searches the prepared image folders for a folder named
\f[I]Fractals\f[R] and copies any images in that folder into the default
backgrounds folder for use as desktop wallpapers and slideshows.
Any previous images in the default backgrounds folder are retained and
the images from the \f[I]Fractals\f[R] folder are added to the default
backgrounds folder.
.SH AUTHORS
.PP
Written by Ronald Record <<github@ronrecord.com>>
.SH LICENSING
.PP
BACKGROUNDS is distributed under an Open Source license.
See the file LICENSE in the BACKGROUNDS source distribution for
information on terms & conditions for accessing and otherwise using
BACKGROUNDS and for a DISCLAIMER OF ALL WARRANTIES.
.SH BUGS
.PP
Submit bug reports online at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts/issues>>
.SH SEE ALSO
.PP
Full documentation and sources at:
<<https://gitlab.com/doctorfree/DoctorFreeScripts>>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

Jan 27, 2022 Version 4.0.1 Release 2
    Minor release update with improved support for SD management utilities,
    extending supported platforms to Linux for the SD utils.

Jan 17, 2022 Version 4.0.1 Release 1
    Release update to Version 4.0.1 including additional man pages
    and Zip format distribution archive. Minor bug fixes and improvements.
    RPM installation fixes.

Nov 26, 2021 Version 4.0 Release 1
    This release adds Debian and RPM format installation package support with
    improved support for ease of installation, configuration, update, and removal.

Wed Jan 26 14:27:43 2022 -0800 b752940 :
   Exec /bin/bash in systemd service to comply with Fedora restrictions
Tue Jan 25 08:37:12 2022 -0800 f38080d :
   Add journalclean command
Tue Jan 25 07:14:52 2022 -0800 d4b92fc :
   Set backgrounds folder to /media/user/Transcend by default on Linux
Mon Jan 24 18:21:57 2022 -0800 363a940 :
   Change default SD device to /dev/sdd
Mon Jan 24 10:17:35 2022 -0800 3be3fab :
   Add documentation for sdmount and sdumount
Sun Jan 23 19:48:26 2022 -0800 4088f8d :
   Add sdmount and sdumount commands
Sun Jan 23 10:31:20 2022 -0800 68b537d :
   Doc for sdlist
Sun Jan 23 10:04:48 2022 -0800 04f8708 :
   Add sdlist command
Sat Jan 22 16:03:14 2022 -0800 02729e7 :
   Extend SD utilities with Linux support
Fri Jan 21 07:50:47 2022 -0800 6be5d58 :
   Muriel aliases
Thu Jan 20 11:45:13 2022 -0800 24d045d :
   Redirect stderr and stdout to parent process during RPM scriptlets
Tue Jan 18 17:17:31 2022 -0800 2152b2d :
   Use DISPLAY env var if set
Tue Jan 18 13:20:58 2022 -0800 85da437 :
   Add doctorinfo script
Mon Jan 17 13:52:51 2022 -0800 65cd49e :
   Adopt release creation during Gitlab CI
Mon Jan 17 13:42:38 2022 -0800 735ff4e :
   Do not invoke dl_fred_scripts during postinstall, new version string
Mon Jan 17 12:06:57 2022 -0800 6bc24c5 :
   Move pvoutput to MirrorCommand
Mon Jan 17 12:06:31 2022 -0800 8dd49ed :
   Bump release
Fri Jan 14 13:52:28 2022 -0800 1ab3135 :
   Update miner scripts
Fri Jan 14 10:49:45 2022 -0800 988389c :
   Add disable-update-notifier and pkg/tools/get_release
Mon Jan 10 12:23:00 2022 -0800 6e4799a :
   Do not package or export gdrive utils
Mon Jan 10 12:19:41 2022 -0800 9c85af0 :
   Remove gdrive files as they are now packaged in their own repo
Sun Jan 9 17:43:49 2022 -0800 8962417 :
   disable rpmbuild strip
Sun Jan 9 17:18:32 2022 -0800 5af1d70 :
   Fix files section of spec file
Sun Jan 9 17:10:20 2022 -0800 905d402 :
   Move all files into /usr/local
Sun Jan 9 16:39:42 2022 -0800 e27d740 :
   Try noninteractive rpm install again
Sun Jan 9 15:35:04 2022 -0800 64ddc3a :
   Fix for CI
Sun Jan 9 15:28:21 2022 -0800 fd5915b :
   export noninteractive
Sun Jan 9 15:18:00 2022 -0800 c6e51e9 :
   Copy rpm dist tree in install phase
Sun Jan 9 14:46:58 2022 -0800 bfa4e07 :
   Install rpm in CI in noninteractive mode with TZ set
Sun Jan 9 14:17:31 2022 -0800 53255f8 :
   Use ubuntu:latest CI image
Sun Jan 9 14:09:05 2022 -0800 c6963e9 :
   Version 4.0 release 3
Sun Jan 9 14:04:52 2022 -0800 fd33ea1 :
   Install zip if not found, create zip artifacts in CI
Thu Jan 6 11:57:39 2022 -0800 3b3f004 :
   Add drive to projects
Mon Jan 3 20:15:25 2022 -0800 748b3e0 :
   Add author attribution
Mon Jan 3 16:22:13 2022 -0800 236f864 :
   Add -l option to list sync folders
Mon Jan 3 16:11:26 2022 -0800 d30988a :
   Renamed getfileids as getfolderids
Mon Jan 3 13:01:47 2022 -0800 97d0e8d :
   Check for gdrive binary in PATH
Mon Jan 3 12:38:50 2022 -0800 ded8f3d :
   Add options to getfolderids, start on support for spaces in file and folder names
Mon Jan 3 10:24:36 2022 -0800 627a90d :
   Support spaces in folder names, retrieve all folder ids
Sun Jan 2 21:10:06 2022 -0800 1ec7cb9 :
   Export gdrive utils
Sun Jan 2 21:07:06 2022 -0800 40e2bdb :
   Add gdrive scripts to exports
Sun Jan 2 21:04:30 2022 -0800 e9d2a92 :
   Add comment for developer's guide to api
Sun Jan 2 20:19:56 2022 -0800 c674bc0 :
   Add sync_from_drive, update sync2drive
Sun Jan 2 11:48:46 2022 -0800 285acd8 :
   Updated gdrive utility scripts
Sat Jan 1 20:14:40 2022 -0800 a670a0d :
   Add new scripts for command line management of Google Drive
Fri Dec 31 08:52:06 2021 -0800 a1659d2 :
   Add Mirco Cabbia alias
Tue Dec 28 09:50:38 2021 -0800 904ab6d :
   Rename MirrorCommandLine to MirrorCommand
Mon Dec 27 18:20:34 2021 -0800 3554d70 :
   New model alias
Mon Dec 27 08:43:35 2021 -0800 5a2a524 :
   New model alias
Mon Dec 27 08:23:33 2021 -0800 a981875 :
   Create man page for md2man
Sun Dec 26 17:48:24 2021 -0800 6d83be7 :
   Set output dir based on man page section in md2man
Sat Dec 25 08:04:35 2021 -0800 ef79abb :
   Added Colombian wallpapers
Fri Dec 24 11:34:02 2021 -0800 0c9e9cc :
   Adjusted aliases
Thu Dec 23 09:38:25 2021 -0800 c2b5191 :
   Updated model name, add Mila Azul to Ukrainians
Wed Dec 22 16:31:35 2021 -0800 d4296b4 :
   Use HEAD to determine branch
Tue Dec 21 14:01:20 2021 -0800 526124a :
   Bump release number, update changelog
Tue Dec 21 10:26:02 2021 -0800 8c3a636 :
   Add all my clone and build scripts
Tue Dec 21 09:43:04 2021 -0800 d625a49 :
   Check and set location of Synergy client binary and system tools in Synergy startup control script
Tue Dec 21 09:14:43 2021 -0800 7a101e1 :
   Add man page for sdwrite, add SEE ALSO pages to all SD card commands
Tue Dec 21 08:40:29 2021 -0800 58b2dfc :
   Move sdwrite from MirrorCommandLine package to here
Sat Dec 18 17:18:51 2021 -0800 148f60a :
   Set synergyctl IP in postinstall
Sat Dec 18 16:39:57 2021 -0800 a8df6ce :
   Add synergyctl script to control Synergy start/stop
Fri Dec 17 07:15:45 2021 -0800 30fd5e5 :
   New model aliases
Mon Dec 13 21:11:37 2021 -0800 2c51ae3 :
   Fix typo in rmbrokenpkg
Mon Dec 13 20:18:02 2021 -0800 e5436b5 :
   Add rmbrokenpkg
Sun Dec 12 13:40:26 2021 -0800 2a4714d :
   Add note about Windows System for Linux

Wallpapers/backgrounds.sh
#!/bin/bash
#
## @file Wallpapers/backgrounds.sh
## @brief Set background desktop wallpapers or display slideshow
## @author Ronald Joe Record (rr at ronrecord dot com)
## @copyright Copyright (c) 2017, Ronald Joe Record, all rights reserved.
## @date Written 17-Sep-2017
## @version 2.0.1
##
## NOTES:
##
## Link the file "backgrounds" to "slides" and/or "slideshow" to enable
## and turn on automatic slideshow presentation when invoked as such.
##
## I found I needed to give iTerm2 system privileges in
## System Preferences -> Security & Privacy -> Privacy
## Adding it to the list of applications given Accessibility and Full Disk Access
## privileges. On another system I needed to add "System Events" rather than "iTerm 2"
## to Accessibility so ymmv - apparently this is a result of strengthened privacy
## protections in Apple's Mojave release.
##
## You may wish to add the "Slideshow" icon to the Preview toolbar for ease of use
##
# My MagicMirror is Linux with a different TOP and PRE
mirror=
[ -d ${HOME}/MagicMirror ] && mirror=1

# Root directory of your subfolders of image files to use as backgrounds/slideshows
# Modify these image folder settings to reflect your image folder layout
if [ "$mirror" ]
then
    USER=`id -u -n`
    # MagicMirror image folders
    LIN_TOP=/media/${USER}/Transcend/Pictures
    PRE_TOP=$HOME/Pictures/Backgrounds
else
    # Linux Desktop image folders
    LIN_TOP=/u/pictures
    PRE_TOP=$HOME/Pictures/Work/Backgrounds
fi
# Mac OS X image folders
MAC_TOP="/Volumes/Seagate_8TB/Pictures/Work"
NFS_TOP=/u/pictures/Work
#
# Location of folder to copy selected images to
MAC_OUT=$HOME/Pictures/Backgrounds
LIN_OUT=/usr/local/share/backgrounds
#
# Location of prepared folders
PRE_DIR=$HOME/Pictures/Work/Backgrounds
#
bak=/tmp/pic$$
maxlinks=2048
add=
all=
osa=
var=
pcman=
paper=
show=
subdir=
foundirs=
updpre=
xnview=
name=`basename $0`
plat=`uname -s`

usage() {
  printf "\nUsage: $name [-u] [-a] [-l] [-x] [-pP] [-S] [-n numpics] [-s subdir] directory"
  printf "\nWhere\t-a indicates add to existing background/slide pics"
  printf "\n\t-l lists currently installed background/slide pics"
  printf "\n\t-p indicates search for background pics in prepared folders"
  printf "\n\t-P indicates search for background pics in prepared folders and update"
  printf "\n\t-n <numpics> sets maximum number of pics to be copied (default ${maxlinks})"
  printf "\n\t-s <subdir> searches in $TOP/<subdir> for specified folder"
  printf "\n\t-x indicates use XnViewMP for slideshow"
  printf "\n\t-S indicates run a slideshow of pics\n\n"
  exit 1
}

# Turn on slideshow capability if invoked as "slides" or "slideshow"
[ "$name" == "slides" ] && show=1
[ "$name" == "slideshow" ] && show=1

# Try to figure out which system we are on
if [ "$plat" == "Linux" ]
then
    TOP="$LIN_TOP"
    OUT="$LIN_OUT"
    SLIDE_APP="Variety Slideshow"
else
    if [ "$plat" == "Darwin" ]
    then
        TOP="$MAC_TOP"
        OUT="$MAC_OUT"
        SLIDE_APP="AppleScript"
    else
        echo "Unable to detect a supported platform: ${plat}. Exiting."
        exit 1
    fi
fi
[ -d "$TOP" ] || {
    TOP="$NFS_TOP"
    [ -d "$TOP" ] || {
        echo "Cannot locate Work directory for pics. Exiting."
        exit 1
    }
}

# Set Model and Photographer top dirs before processing arguments
MOD_TOP="$TOP/Wallhaven/Models"
PHO_TOP="$TOP/Wallhaven/Photographers"
  
while getopts pPn:s:ab:lxSu flag; do
    case $flag in
        a)
            add=1
            ;;
        b)
            paper="$OPTARG"
            ;;
        l)
#           ls --color=auto -l $OUT | awk ' { print $NF } '
            numprep=`ls -1 $OUT | wc -l`
            ls -l $OUT | grep -v total | awk ' { print $NF } '
            echo "$numprep background images in $OUT"
            echo ""
            echo "Prepared folders in $PRE_DIR :"
            for folder in $PRE_DIR/*
            do
                [ "$folder" == "$PRE_DIR/*" ] && continue
                [ -d "$folder" ] && {
                    prepmod=`basename "$folder"`
                    echo "$prepmod"
                }
            done
            exit 0
            ;;
        n)
            maxlinks="$OPTARG"
            ;;
        p)
            TOP=$PRE_TOP
            foundirs="$TOP"
            ;;
        P)
            TOP=$PRE_TOP
            foundirs="$TOP"
            updpre=1
            ;;
        s)
            case "$OPTARG" in
              all) all=1
                ;;
              elite) subdir="Elite_Babes"
                ;;
              fractals) subdir="Fractals"
                ;;
              jp) subdir="JP_Erotica"
                ;;
              kind) subdir="KindGirls"
                ;;
              whvn) subdir="Wallhaven/Models"
                ;;
              xart) subdir="X-Art"
                ;;
              *) subdir="$OPTARG"
                ;;
            esac
            ;;
        x)
            xnview=1
            ;;
        S)
            show=1
            ;;
        u)
            usage
            ;;
    esac
done
shift $(( OPTIND - 1 ))

foundpaper=
if [ "$paper" ]
then
#   echo "Checking for $paper"
    if [ -f "$paper" ]
    then
      foundpaper=1
    else
#     echo "Checking $OUT"
      if [ -f "${OUT}/$paper" ]
      then
          paper="${OUT}/$paper"
          foundpaper=1
      else
#         echo "Checking $TOP"
          if [ -f "${TOP}/$paper" ]
          then
              paper="${TOP}/$paper"
              foundpaper=1
          else
#             echo "Checking $PRE_TOP"
              if [ -f "${PRE_TOP}/$paper" ]
              then
                  paper="${PRE_TOP}/$paper"
                  foundpaper=1
              else
                  echo "Cannot find specified background pic $paper"
              fi
          fi
      fi
    fi
    # TODO: Set background on systems without pcmanfm
    [ "$foundpaper" ] && DISPLAY=${DISPLAY:=:0} pcmanfm --set-wallpaper=$paper
fi

[ "$show" ] && {
    if [ "$plat" == "Darwin" ]
    then
        inst=`type -p osascript`
        [ "$inst" ] && osa=1
    else
        inst=`type -p variety-slideshow`
        [ "$inst" ] && var=1
        inst=`type -p pcmanslideshow`
        [ "$inst" ] && pcman=1
    fi
    [ "$inst" ] || {
        echo "$SLIDE_APP is not supported on this platform."
        echo "Unable to automate the slideshow feature."
    }
}

[ -d $OUT ] || {
    sudo mkdir $OUT
    user=`id -u -n`
    sudo chown $user $OUT
}

cd $OUT

[ "$1" ] && {
  [ "$add" ] || {
    mkdir $bak
    touch $bak/foo
    for i in *
    do
      [ "$i" == "*" ] && continue
      mv $i $bak
    done
  }

  bdir="$1"
  if [ "$bdir" == "favs" ]
  then
    TOP="$HOME/Pictures"
    bdir="Favs"
    foundirs=$TOP
  else
    if [ "$subdir" ]
    then
      [ -d "$TOP/$subdir/$bdir" ] || {
        echo "Cannot locate $TOP/$subdir/$bdir - exiting."
        exit 1
      }
      TOP="$TOP/$subdir"
      foundirs="$TOP"
    else
      [ -d "$TOP/$bdir" ] || {
        [ "$updpre" ] && mkdir "$TOP/$bdir"
        for subdir in Wallhaven Wallhaven/Models Wallhaven/Photographers X-Art Elite_Babes JP_Erotica Met-Art KindGirls Wallbase
        do
          [ -d "$TOP/$subdir/$bdir" ] && {
            sub="$TOP/$subdir"
            if [ "$all" ]
            then
              foundirs="$foundirs $sub"
            else
              foundirs="$sub"
              break
            fi
          }
        done
        #TOP="$sub"
      }
    fi
  fi
  [ -d "$TOP/$bdir" ] || {
    echo "Cannot locate $TOP/$bdir - exiting."
    exit 1
  }
  echo "Using background pics in $TOP/$bdir"

  numlinks=0
  for dir in $foundirs
  do
    [ "$updpre" ] && {
      updtop=
      if [ -d $MOD_TOP/$bdir ]
      then
        updtop=$MOD_TOP
      else
        if [ -d $PHO_TOP/$bdir ]
        then
          updtop=$PHO_TOP
        fi
      fi
      [ "$updtop" ] && {
        [ -d $TOP/$bdir ] || mkdir $TOP/$bdir
        cd $TOP/$bdir
        for orig in $updtop/$bdir/wallhaven*
        do
          [ "$orig" == "$updtop/$bdir/wallhaven*" ] && {
            echo "No pics in $updtop/$bdir"
            continue
          }
          bnam=`basename $orig`
          [ -L $bnam ] && continue
          cp $orig $bnam
          rmportargs $bnam
          [ -f $bnam ] && {
            rm -f $bnam
            echo "Adding $bnam to $bdir"
            ln -s $orig .
          }
        done
        cd $OUT
      }
    }
    for pic in $dir/$bdir/*
    do
      [ "$pic" == "$dir/$bdir/*" ] && continue
      [ "$pic" == "$dir/$bdir/downloaded.txt" ] && continue
      [ "$pic" == "$dir/$bdir/Description.txt" ] && continue
      if [ -d "$pic" ]
      then
        for subpic in $pic/*
        do
          [ "$subpic" == "$pic/*" ] && continue
          [ "$subpic" == "$pic/downloaded.txt" ] && continue
          [ "$subpic" == "$pic/Description.txt" ] && continue
          [ -d "$subpic" ] && continue
#       bnam=`basename $subpic`
#       [ -L $bnam ] && continue
#       ln -s $subpic .
          cp "$subpic" .
          numlinks=`expr $numlinks + 1`
          [ $numlinks -ge $maxlinks ] && break 2
        done
      else
#     bnam=`basename $pic`
#     [ -L $bnam ] && continue
#     ln -s $pic .
        cp "$pic" .
        numlinks=`expr $numlinks + 1`
      fi
      [ $numlinks -ge $maxlinks ] && break 2
    done
  done

  for j in *
  do
    [ "$j" == "*" ] && {
      [ "$add" ] || {
        mv $bak/* .
        rm -f foo
      }
      continue
    }
    file -L "$j" | grep ASCII > /dev/null && rm -f "$j"
  done
  [ "$add" ] || rm -rf $bak
}

[ "$show" ] && {
    if [ "$plat" == "Darwin" ]
    then
        if [ "$xnview" ]
        then
            open -a XnViewMP "$OUT"
        else
            open -a Preview "$OUT"/*
        fi
        # ----------------------
        # Applescript below here
        # ----------------------
        [ "$osa" ] && {
          if [ "$xnview" ]
          then
            osascript <<EOF
            delay 5
            tell application "System Events"
                keystroke "a" using {command down}
                delay 3
                key code 36
                delay 3
                keystroke "s" using {option down, command down}
                delay 3
                key code 36
                delay 3
                keystroke "f" using {option down, command down}
            end tell
EOF
          else
            osascript <<EOF
            delay 5
            tell application "System Events"
                keystroke "F" using {shift down, command down}
            end tell
EOF
          fi
        }
    else
#       variety-slideshow $HOME/.config/variety/Favorites 2> /dev/null
        if [ "${var}" ]
        then
            variety-slideshow "$OUT" 2> /dev/null
        else
            if [ "${pcman}" ]
            then
                DISPLAY=${DISPLAY:=:0} pcmanslideshow
            else
                echo "No slideshow program found for this platform"
            fi
        fi
    fi
}
